### PR TITLE
Test coverage

### PR DESCRIPTION
**Pull request by RedEnchilada**
Add tests for a lot of the functionality currently exposed by FlexibleMink. Not everything is covered yet, but a lot of commonly-used functionality now has tests ensuring that `Then`s both pass and fail (in all cases) properly, or that `When`s produce the expected action. (This means the Kanban story isn't quite complete yet, but I figured it'd be better to get early feedback in this process instead of bunching it all up into one monstrous PR after doing everything.)

A few fixes were also included to contexts where undesired behavior was discovered while writing tests.

### CODE REVIEW

**Comment by medology-bugman** [comment]
Paging @Chekote to review the back-end work in this pull request!
Paging @johnnie502 to review the testing work in this pull request!
Paging @TaysirTayyab to review the front-end work in this pull request!

I couldn't recognize who these files were relevant to:
behat.yml
bin/php

**Comment by TaysirTayyab** [comment]
This is absolutely fantastic @RedEnchilada! Front-end review passed a888724 👍

File: features/StoreContext.feature [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+     Then the assertion should pass
+
+    Examples:
+      | Value |
+      | 2     |
+      | 0     |
+      | 4.7   |
+      | true  |
+      | false |
+      | "car" |
```
Comment (johnnie502) [inline features/StoreContext.feature]: We may need to assert if class instance could be stored and accessed

**Comment by johnnie502** [comment]
Testing review complete, this is almost a comprehensive test set for flexible-mink, good job!

File: bin/php [hunk 2]
```diff
@@ -5,4 +5,4 @@ for i in "$@"; do
   ARGS="$ARGS ${i}"
 done
 
-docker run -it -v $(pwd):/app -w /app php:cli php $ARGS
+docker run -it -v $(pwd):/app --net flexiblemink_default --link chrome --rm -w /app php:cli php $ARGS
```
Comment (Chekote) [inline bin/php]: I believe you can access all containers in the network without specifying the link argument.

File: features/AlertContext.feature [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      | confirm |
+      | cancel  |
+
+  Scenario: Alert assertion fails properly when no alert exists
+    Given I am on "index.html"
+     When I assert that I should see an alert containing "alert text"
+     Then the assertion should throw an ExpectationException
+      And the assertion should fail with the message "No alert is open"
+
+  Scenario: Alert assertion fails properly when no alert exists
```
Comment (Chekote) [inline features/AlertContext.feature]: This Scenario name doesn't seem correct. The same file is being opened on line 7 of this Feature and the assertion is passing. So an alert does exist. It just does not match the specified text.

- [ ] Done
- [ ] Approved

File: features/AlertContext.feature [hunk 4]
```diff
@@ -0,0 +1,43 @@
+Feature: Alert context
+  In order to ensure that JavaScript alerts work as expected
+  As a developer
+  I need to be able to assert various states of alerts
+
+  Scenario Outline: Developer can test when alerts appear
+    Given I am on "alert.html"
```
Comment (Chekote) [inline features/AlertContext.feature]: If we encapsulate this in a step that describes the intended behavior, then this Scenario would be easier to read and not quite so "coincidental":

```cucumber
Scenario Outline: Developer can test when alerts appear
  Given there is an alert containing "alert text"
   When I assert that I should see an alert containing "alert text"
  ...
```

- [ ] Done
- [ ] Approved

File: features/AlertContext.feature [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+     When I <Action> the alert
+     Then I should see "<Action> selected"
+
+    Examples:
+      | Action  |
+      | confirm |
+      | cancel  |
+
+  Scenario: Alert assertion fails properly when no alert exists
+    Given I am on "index.html"
```
Comment (Chekote) [inline features/AlertContext.feature]: Same thing here with describing the intended state:

```cucumber
Scenario Outline: Alert assertion fails properly when no alert exists
  Given there is not an alert containing "alert text"
   When I assert that I should see an alert containing "alert text"
  ...
```

- [ ] Done
- [ ] Approved

File: features/AlertContext.feature [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+     When I assert that I should see an alert containing "alert text"
+     Then the assertion should throw an ExpectationException
+      And the assertion should fail with the message "No alert is open"
+
+  Scenario: Alert assertion fails properly when no alert exists
+    Given I am on "alert.html"
```
Comment (Chekote) [inline features/AlertContext.feature]: Same thing here with describing the intended state:

```cucumber
Scenario Outline: Alert assertion fails properly when no alert exists
  Given there is an alert containing "alert text"
   When I assert that I should see an alert containing "alert butt"
  ...
```

- [ ] Done
- [ ] Approved

File: features/AlertContext.feature [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+     When I assert that I should see an alert containing "alert butt"
+     Then the assertion should throw an ExpectationException
+      And the assertion should fail with the message "Text 'alert butt' not found in alert"
+          # Selenium bombs out if we leave an alert up at the end of a test
+     When I confirm the alert
```
Comment (Chekote) [inline features/AlertContext.feature]: You should be able to remove this and instead add the following to the context:

```php
    /**
     * Restarts the browser session.
     *
     * @BeforeScenario
     */
    public function restartSession()
    {
        $this->getSession()->restart();
    }
```

- [ ] Done
- [ ] Approved

File: features/AlertContext.feature [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+  Scenario Outline: Developer can test the results of filling in prompts
+    Given I am on "prompt.html"
```
Comment (Chekote) [inline features/AlertContext.feature]: Same thing here with describing the intended state:

```cucumber
Scenario Outline: Developer can test the results of filling in prompts
  Given there is a prompt containing "prompt?"
   When I assert that I should see a prompt containing "prompt?"
  ...
```

- [ ] Done
- [ ] Approved

**Comment by Chekote** [comment]
Back end review for a888724 is complete.

Great work on filling out the tests for this project!

File: features/AlertContext.feature (hunk 7 above)
Comment (RedEnchilada) [inline features/AlertContext.feature]: I added an `@AfterScenario` hook to `AlertContext` that closes any open alerts instead. It has an additional benefit of preventing the whole testing suite from bombing out if another project using FlexibleMink runs into an unexpected alert.

File: features/StoreContext.feature (hunk 1 above)
Comment (RedEnchilada) [inline features/StoreContext.feature]: There's no code in `StoreContext` to handle classes in particular, so what would be tested?

Comment (johnnie502) [inline features/StoreContext.feature]: All right, seems it doesn't matter then.

File: features/AlertContext.feature (hunk 7 above)
Comment (Chekote) [inline features/AlertContext.feature]: Does the `$this->getSession()->restart();` not work if there are alerts present?

**Comment by Chekote** [comment]
Assuming that it is not possible to restart a session if there are alerts (see [this comment](https://github.com/medology/flexiblemink/pull/58#discussion_r75300210)), then this is good to merge.

3dfc7cb 👍

**Comment by johnnie502** [comment]
👍  Testing

File: features/AlertContext.feature (hunk 7 above)
Comment (RedEnchilada) [inline features/AlertContext.feature]: I went with closing the alert for performance reasons. Using `$this->getSession()->restart();` takes the tests from 0m49.95s to 1m34.12s (almost a 1s increase per scenario), which seems unreasonable for tests that gain nothing else out of the session reboot. Using`$this->getSession()->reset();` instead doesn't solve the issue of alerts staying open, so it still bombs out.

**Pull request merged**